utils/lsa/chunked_managed_vector.hh

```diff
@@ -26,16 +26,17 @@ namespace lsa {
 
 template <typename T>
 class chunked_managed_vector {
-    static const size_t max_contiguous_allocation = logalloc::segment_size * 0.1;
-    static_assert(std::is_nothrow_move_constructible<T>::value, "T must be nothrow move constructible");
     using chunk_ptr = managed_vector<T>;
+    static constexpr size_t max_contiguous_allocation = logalloc::max_managed_object_size - chunk_ptr::metadata_size();
```

Comment:
It would be better to make it depend on the actual allocator rather than be tied to logalloc. This would make transient chunked_managed_vectors more efficient. But I don't propose changing it here.

Reply:
We don't want transient chunked_managed_vectors. For transient uses, there's chunked_vector.

Reply:
We could have them for data structures that are shared between lsa and non-lsa (it's a different question whether those are a good idea or not).

We only have one user, partition_index_range.

```c++
// A partition index page.
//
// Allocated in the standard allocator space but with an LSA allocator as the current allocator.
// So the shallow part is in the standard allocator but all indirect objects are inside LSA.
class partition_index_page {
public:
    lsa::chunked_managed_vector<managed_ref<index_entry>> _entries;
```

Does it even need to be managed, given the comment?

Reply:
> Does it even need to be managed

Yes.

>  given the comment

The comment is only saying that `partition_index_page` is allocated in the standard allocator, but the contents of `partition_index_page::_entries` (and their children) are allocated in LSA.

Reply:
If partition_index_page is transient, why does it make sense to allocate the chunked_managed_vector contents in lsa? It just generates garbage to be collected.

Why not a chunked_vector\<managed_ref\<index_entry>>? It allows the index_entries to be persistent in lsa, while the page itself is transient.

Reply:
> If partition_index_page is transient, why does it make sense to allocate the chunked_managed_vector contents in lsa? It just generates garbage to be collected.

It's not transient all the time. partition_index_page is allocated in the std allocator during loading, but then it's moved into cache entry and the whole thing becomes managed by LSA. _entries are always managed in LSA, even during loading.

Reply:
Aha. So it's allocated outside lsa, but with anticipation of being moved into lsa, so https://github.com/scylladb/scylladb/pull/23944#discussion_r2066222797 wouldn't help here.

Still, the comment is wrong or at least misleading.

Reply:
> It would be better to make it depend on the actual allocator rather than be tied to logalloc. This would make transient chunked_managed_vectors more efficient. But I don't propose changing it here.

Yes, managed containers should respect preferred_max_contiguous_allocation().

> We only have one user, partition_index_range.

So it's ok now, but later we may add another user and no one will notice this detail. I think we should follow general principles when implementing general-purpose containers.

Reply:
> I think we should follow general principles when implementing general-purpose containers.

Let's check who hardcoded that size originally: 78e5b9fd8542a1621f6e49cd368725e089813629

Reply:
> > I think we should follow general principles when implementing general-purpose containers.
> 
> Let's check who hardcoded that size originally: [78e5b9f](https://github.com/scylladb/scylladb/commit/78e5b9fd8542a1621f6e49cd368725e089813629)

Well, that doesn't invalidate the point.

But now I remember why it's like this: we need stable addr(), which doesn't have to be invoked in allocation context, so we either we make the chunk size static like now, or store as member field at allocation time, which bloats the size of the object. So perhaps the current trade-off is fine.

Reply:
Anyway, not introduced here so not a blocker.